simple_slice.py

```diff
@@ -0,0 +1,294 @@
+    # numpy uses __getitem__ rather than __iter__ to index into shape, so we
+    # match that
+    for i in range(l):
+        # Raise TypeError if invalid
+        val = shape[i]
+        if val is None:
+            raise ValueError("unknonwn (None) dimensions are not supported")
+
+        newshape.append(operator_index(shape[i]))
+
+        if not allow_negative and val < 0:
+            raise ValueError("unknown (negative) dimensions are not supported")
+
+    if axis is not None:
+        if len(newshape) <= axis:
+            raise IndexError(f"too many indices for array: array is {len(shape)}-dimensional, but {axis + 1} were indexed")
+
+    return tuple(newshape)
+
+def operator_index(idx):
+    if isinstance(idx, bool):
+        raise TypeError("'bool' object cannot be interpreted as an integer")
+    if 'numpy' in sys.modules and isinstance(idx, sys.modules['numpy'].bool_):
+        raise TypeError("'np.bool_' object cannot be interpreted as an integer")
+    return operator.index(idx)
+
+class default:
+    pass
+
+class ImmutableObject:
+    __slots__ = ('args',)
+
+    def __init__(self, *args, **kwargs):
+        args = self._typecheck(*args, **kwargs)
+        self.args = args
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({', '.join(map(repr, self.args))})"
+
+    def __str__(self):
+        return f"{self.__class__.__name__}({', '.join(map(str, self.args))})"
+
+    def __eq__(self, other):
+        if not isinstance(other, ImmutableObject):
+            try:
+                other = self.__class__(other)
+            except TypeError:
+                return False
+
+        return self.args == other.args
+
+class SimpleSlice(ImmutableObject):
+    __slots__ = ()
+
+    def _typecheck(self, start, stop=default, step=None):
+        if isinstance(start, SimpleSlice):
+            return start.args
+        if isinstance(start, slice):
+            start, stop, step = start.start, start.stop, start.step
+
+        if stop is default:
+            start, stop = None, start
+
+        if step == 0:
+            raise ValueError("slice step cannot be zero")
+
+        if start is not None:
+            start = operator_index(start)
+        if stop is not None:
+            stop = operator_index(stop)
+        if step is not None:
+            step = operator_index(step)
+
+        args = (start, stop, step)
+
+        return args
+
+    @property
+    def raw(self):
+        return slice(*self.args)
+
+    @property
+    def start(self):
+        return self.args[0]
+
+    @property
+    def stop(self):
+        return self.args[1]
+
+    @property
+    def step(self):
+        return self.args[2]
+
+class SimpleSliceSubclass(SimpleSlice):
+    pass
+
+class SimpleSliceCythonSubclass(SimpleSliceCython):
+    pass
+
+class SimpleSlicePybind11Subclass(SimpleSlicePybind11):
+    pass
+
+class SimpleSliceRustSubclass(SimpleSliceRust):
+    pass
+
+def reduce(self, shape=None, *, axis=0, negative_int=False):
```

Comment:
This is a big offender in the slowness of `versioned_hdf5.wrappers.as_subchunk_map`.
I managed to halve the end-to-end runtime by omitting it entirely and rewriting a special case of it in cython:
https://github.com/deshaw/versioned-hdf5/pull/364#discussion_r1719080715